File: MicroTDCore/Sources/MicroTDCore/Systems/RunStateMachine.swift
// RunStateMachine.swift
// Enforces valid state transitions for run lifecycle

/// Run lifecycle states
/// Enforces orderly progression through game phases
public enum RunState: Equatable {
    case preRun
    case building(waveIndex: Int)
    case inWave(waveIndex: Int)
    case relicChoice(waveIndex: Int)
    case gameOver
    case postRunSummary
}

/// State transition errors
public enum StateTransitionError: Error {
    case invalidTransition(from: RunState, to: RunState)
    case waveIndexMismatch
}

/// Manages run state with enforced transitions
public struct RunStateMachine {
    public private(set) var state: RunState
    
    public init(initialState: RunState = .preRun) {
        self.state = initialState
    }
    
    /// Attempt to transition to new state, throws if invalid
    public mutating func transition(to newState: RunState) throws {
        guard isValidTransition(from: state, to: newState) else {
            throw StateTransitionError.invalidTransition(from: state, to: newState)
        }
        state = newState
    }
    
    private func isValidTransition(from: RunState, to: RunState) -> Bool {
        switch (from, to) {
        // From preRun
        case (.preRun, .building(let wave)):
            return wave == 0
            
        // From building
        case (.building(let wave1), .inWave(let wave2)):
            return wave1 == wave2
            
        // From inWave
        case (.inWave(let wave1), .building(let wave2)):
            return wave2 == wave1 + 1
        case (.inWave, .relicChoice):
            return true
        case (.inWave, .gameOver):
            return true
        case (.inWave, .postRunSummary):
            return true
            
        // From relicChoice
        case (.relicChoice(let wave1), .building(let wave2)):
            return wave2 == wave1 + 1
            
        // From gameOver (revive support)
        case (.gameOver, .inWave):
            return true
        case (.gameOver, .postRunSummary):
            return true
            
        // From postRunSummary (end of run)
        case (.postRunSummary, .preRun):
            return true
            
        default:
            return false
        }
    }
}

File: expo/src/core/systems/RunStateMachine.ts
// RunStateMachine.ts
// Game state lifecycle — ported from MicroTDCore/Systems/RunStateMachine.swift

/**
 * All possible run states
 */
export type RunStateType = 'preRun' | 'building' | 'inWave' | 'relicChoice' | 'gameOver' | 'postRunSummary';

/**
 * Run state with associated data
 */
export type RunState =
    | { type: 'preRun' }
    | { type: 'building'; waveIndex: number }
    | { type: 'inWave'; waveIndex: number }
    | { type: 'relicChoice'; waveIndex: number }
    | { type: 'gameOver'; wavesCompleted: number; didWin: boolean }
    | { type: 'postRunSummary'; wavesCompleted: number; didWin: boolean };

/**
 * Valid state transitions
 */
const VALID_TRANSITIONS: Record<RunStateType, RunStateType[]> = {
    preRun: ['building'],
    building: ['inWave'],
    inWave: ['building', 'relicChoice', 'gameOver'],
    relicChoice: ['building'],
    gameOver: ['postRunSummary'],
    postRunSummary: ['preRun'],
};

/**
 * Manages the lifecycle state of a game run.
 * Enforces valid transitions and prevents invalid state changes.
 */
export class RunStateMachine {
    private _state: RunState;

    constructor() {
        this._state = { type: 'preRun' };
    }

    get state(): RunState {
        return this._state;
    }

    /**
     * Transition to a new state.
     * Throws if the transition is not valid from the current state.
     */
    transition(to: RunState): void {
        const validTargets = VALID_TRANSITIONS[this._state.type];
        if (!validTargets.includes(to.type)) {
            throw new Error(
                `Invalid state transition: ${this._state.type} → ${to.type}. ` +
                `Valid transitions from ${this._state.type}: ${validTargets.join(', ')}`
            );
        }
        this._state = to;
    }

    /**
     * Check if a transition is valid without performing it
     */
    canTransition(to: RunStateType): boolean {
        return VALID_TRANSITIONS[this._state.type].includes(to);
    }

    /**
     * Convenience: start the run (preRun → building)
     */
    startRun(): void {
        this.transition({ type: 'building', waveIndex: 0 });
    }
}
